### PR TITLE
Snapshot queue engine

### PR DESCRIPTION
**Pull request by cef62**
This PR allow `snapguidist` to have control over how many request are sent to the server. As additional bonus after UI redesign triggered by hot-reload only changed snapshot will be sent to the server to be tested.

### CODE REVIEW

File: test/context/testQueueFactory.spec.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    }
+  }
+)
+
+jest.mock(
+  '../../src/context/runTest',
+  () => jest.fn(() => Promise.resolve({ pass: true }))
+)
+
+function delay(time = 0) {
```
Comment (MicheleBertoli) [inline test/context/testQueueFactory.spec.js]: Couldn't this be just `Promise.resolve()`?

File: test/context/snapguidistContextFactory.spec.js [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+  expect(typeof runTest).toBe('function')
+
+  expect(Object.keys(context).length).toBe(3)
+})
+
+test('should add requested snapshot to return true', () => {
+  const { runTest } = snapguidistContextFactory()
+
+  const result = runTest('name', 'reactElement')
+  expect(result).toBeTruthy()
```
Comment (MicheleBertoli) [inline test/context/snapguidistContextFactory.spec.js]: This seems a bit weak.

File: src/context/testQueueFactory.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+          runNextTest()
+        }
+
+        queue.delete(name)
+
+        running.set(
+          name,
+          runTest(snapshot, update)
+          .then((response) => {
+            notify({ response, name, snapshot, update })
```
Comment (MicheleBertoli) [inline src/context/testQueueFactory.js]: The param object seems to have some unnecessary attributes.

File: src/components/Preview.js [hunk 4]
```diff
@@ -24,34 +34,50 @@ class SnapguidistPreview extends Component {
...
+    }
+  }
+
+  runTest(update) {
+    const {
+      name,
+      snapguidist: { runTest } = {},
+    } = this.context
+
+    const { isFetching, response } = runTest(name, this.example, update)
```
Comment (MicheleBertoli) [inline src/components/Preview.js]: Some thoughts:
1. I'm not sure about giving the full responsibility of managing the state to `runTest`.
2. `isQueuing` seems more appropriate than `isFetching`.
3. we should avoid setting the state when the values aren't changed.

File: test/context/testQueueFactory.spec.js (hunk 1 above)
Comment (cef62) [inline test/context/testQueueFactory.spec.js]: The mock promise return the object to emulate the expected result and have test on line 72 to behave as expected.
Test name is : 'should notify listeners when a test return its result'. I can modify the test to reflect an empty value returned from the promise but I'm not convinced.

File: test/context/snapguidistContextFactory.spec.js (hunk 2 above)
Comment (cef62) [inline test/context/snapguidistContextFactory.spec.js]: You're right, also the test is not updated to the actual value returned by `runtTest`, I'll update it

File: src/context/testQueueFactory.js (hunk 3 above)
Comment (cef62) [inline src/context/testQueueFactory.js]: Correct, they are there from previous implementation requirements, I'll remove them

File: src/components/Preview.js (hunk 4 above)
Comment (cef62) [inline src/components/Preview.js]: 1. I'm not sure how to change that, also in the evolution of the feature this part changes a little. If you agree I'll wait for the other WIP features to re-discuss that point.
2. I liked it! I'll refactor it.
3. Totally agree, I'll change the code

**Comment by cef62** [comment]
Thanks @MicheleBertoli, I've fixed most of the inline comments you added.

For the more general comments:

> What about renaming the "context" folder to "queue"?

I like that but, if it's ok for you I'd like to do it after we close the other features in development: `snapguidist-configuration` and `batched-queue`. Mostly to avoid painful rebase sessions ;) 

> I still believe the comments are too much, at least on the parameters :)

No problem on shorten them, but I'm not sure on what comments are you referring to ;)

> The tests of the factories seem a bit redundant, I would suggest you to isolate each layer (e.g. they are all testing that the listeners get notified).

I agree tests can be further improved, as for the other point I'll prefer do it after we close the other 2 feature branches, a lot of tests changed and rebasing them could be longer than it's worth it..  Anyway if you prefer to do it here anyway i'll do it 👍 , let me know.

File: test/context/testQueueFactory.spec.js (hunk 1 above)
Comment (MicheleBertoli) [inline test/context/testQueueFactory.spec.js]: I'm talking about the `delay` function, which doesn't seem to return anything.
Am I wrong?

**Comment by MicheleBertoli** [comment]
re: comments

I believe this is too much :)
(especially if we do it for just a couple of functions)

```java
* @param {string} name - Unique name of the component to be tested
* @param {ReactElement} component - ReactElement of the component to be tested
* @param {boolean} [update] - It specifies if the test must check or update the stored snapshot
* @return {boolean} Returns true if the test is run, false otherwise
```

---

re: merge vs rebase

If you feel happy with this PR, please merge it.
I'm going to approve the changes right now.
I might work on some refactoring over the weekend.

File: test/context/testQueueFactory.spec.js (hunk 1 above)
Comment (cef62) [inline test/context/testQueueFactory.spec.js]: Sorry I misunderstood the method you were referring... 
For the `delay` method you are right ;) I've updated the code!

**Comment by cef62** [comment]
Thanks @MicheleBertoli ! I've fixed the remaining issues. I've left out the renaming of `context` to `queue` and the reorganization of the tests, as already said I'll address them when the other active PRs will be merged.

**Pull request merged**